File: Source/AlphaTab/Model/Voice.cs
using AlphaTab.Audio;
using AlphaTab.Collections;

namespace AlphaTab.Model
{
    /// <summary>
    /// A voice represents a group of beats 
    /// that can be played during a bar. 
    /// </summary>
    public class Voice
    {
        private FastDictionary<int, Beat> _beatLookup;

        /// <summary>
        /// Gets or sets the zero-based index of this voice within the bar. 
        /// </summary>
        public int Index { get; set; }
        /// <summary>
        /// Gets or sets the reference to the bar this voice belongs to. 
        /// </summary>
        public Bar Bar { get; set; }

        /// <summary>
        /// Gets or sets the list of beats contained in this voice. 
        /// </summary>
        public FastList<Beat> Beats { get; set; }

        /// <summary>
        /// Gets or sets a value indicating whether this voice is empty. 
        /// </summary>
        public bool IsEmpty
        {
            get; set;
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="Voice"/> class.
        /// </summary>
        public Voice()
        {
            Beats = new FastList<Beat>();
            IsEmpty = true;
        }

        internal static void CopyTo(Voice src, Voice dst)
        {
            dst.Index = src.Index;
            dst.IsEmpty = src.IsEmpty;
        }

        internal void InsertBeat(Beat after, Beat newBeat)
        {
            newBeat.NextBeat = after.NextBeat;
            if (newBeat.NextBeat != null)
            {
                newBeat.NextBeat.PreviousBeat = newBeat;
            }
            newBeat.PreviousBeat = after;
            newBeat.Voice = this;
            after.NextBeat = newBeat;
            Beats.InsertAt(after.Index + 1, newBeat);
        }


        internal void AddBeat(Beat beat)
        {
            beat.Voice = this;
            beat.Index = Beats.Count;
            Beats.Add(beat);
            if (!beat.IsEmpty)
            {
                IsEmpty = false;
            }
        }

        private void Chain(Beat beat)
        {
            if (Bar == null) return;

            if (beat.Index < Beats.Count - 1)
            {
                beat.NextBeat = Beats[beat.Index + 1];
                beat.NextBeat.PreviousBeat = beat;
            }
            else if (beat.IsLastOfVoice && beat.Voice.Bar.NextBar != null)
            {
                var nextVoice = Bar.NextBar.Voices[Index];
                if (nextVoice.Beats.Count > 0)
                {
                    beat.NextBeat = nextVoice.Beats[0];
                    beat.NextBeat.PreviousBeat = beat;
                }
                else
                {
                    beat.NextBeat.PreviousBeat = beat;
                }
            }
        }

        internal void AddGraceBeat(Beat beat)
        {
            if (Beats.Count == 0)
            {
                AddBeat(beat);
                return;
            }

            // remove last beat
            var lastBeat = Beats[Beats.Count - 1];
            Beats.RemoveAt(Beats.Count - 1);

            // insert grace beat
            AddBeat(beat);
            // reinsert last beat
            AddBeat(lastBeat);

            IsEmpty = false;
        }

        internal Beat GetBeatAtDisplayStart(int displayStart)
        {
            if (_beatLookup.ContainsKey(displayStart))
            {
                return _beatLookup[displayStart];
            }

            return null;
        }

        internal void Finish(Settings settings)
        {
            _beatLookup = new FastDictionary<int, Beat>();
            for (var index = 0; index < Beats.Count; index++)
            {
                var beat = Beats[index];
                beat.Index = index;
                Chain(beat);
            }

            var currentDisplayTick = 0;
            var currentPlaybackTick = 0;
            for (var i = 0; i < Beats.Count; i++)
            {
                var beat = Beats[i];
                beat.Index = i;
                beat.Finish(settings);

                if (beat.GraceType == GraceType.None || beat.GraceType == GraceType.BendGrace)
                {
                    beat.DisplayStart = currentDisplayTick;
                    beat.PlaybackStart = currentPlaybackTick;
                    currentDisplayTick += beat.DisplayDuration;
                    currentPlaybackTick += beat.PlaybackDuration;
                }
                else
                {
                    if (beat.PreviousBeat == null || beat.PreviousBeat.GraceType == GraceType.None)
                    {
                        // find note which is not a grace note
                        Beat nonGrace = beat;
                        int numberOfGraceBeats = 0;
                        while (nonGrace != null && nonGrace.GraceType != GraceType.None)
                        {
                            nonGrace = nonGrace.NextBeat;
                            numberOfGraceBeats++;
                        }

                        var graceDuration = Duration.Eighth;
                        int stolenDuration = 0;
                        if (numberOfGraceBeats == 1)
                        {
                            graceDuration = Duration.Eighth;
                        }
                        else if (numberOfGraceBeats == 2)
                        {
                            graceDuration = Duration.Sixteenth;
                        }
                        else
                        {
                            graceDuration = Duration.ThirtySecond;
                        }

                        if (nonGrace != null)
                        {
                            nonGrace.UpdateDurations();
                        }

                        // grace beats have 1/4 size of the non grace beat following them
                        var perGraceDuration = nonGrace == null
                            ? Duration.ThirtySecond.ToTicks()
                            : (nonGrace.DisplayDuration / 4) / numberOfGraceBeats;

                        // move all grace beats 
                        var graceBeat = Beats[i];
                        for (int j = 0; j < numberOfGraceBeats && graceBeat != null; j++)
                        {
                            graceBeat.Duration = graceDuration;
                            graceBeat.UpdateDurations();

                            graceBeat.DisplayStart = currentDisplayTick - (numberOfGraceBeats - j + 1) * perGraceDuration;
                            graceBeat.DisplayDuration = perGraceDuration;

                            stolenDuration += graceBeat.PlaybackDuration;

                            graceBeat = graceBeat.NextBeat;
                        }

                        // steal needed duration from beat duration
                        if (beat.GraceType == GraceType.BeforeBeat)
                        {
                            if (beat.PreviousBeat != null)
                            {
                                beat.PreviousBeat.PlaybackDuration -= stolenDuration;
                            }
                            currentPlaybackTick -= stolenDuration;
                        }
                        else if (nonGrace != null && beat.GraceType == GraceType.OnBeat)
                        {
                            nonGrace.PlaybackDuration -= stolenDuration;
                        }
                    }

                    beat.PlaybackStart = currentPlaybackTick;
                    currentPlaybackTick = beat.PlaybackStart + beat.PlaybackDuration;
                }
                beat.FinishTuplet();
                _beatLookup[beat.DisplayStart] = beat;
            }
        }

        public int CalculateDuration()
        {
            if (IsEmpty || Beats.Count == 0)
            {
                return 0;
            }

            var lastBeat = Beats[Beats.Count - 1];
            return lastBeat.PlaybackStart + lastBeat.PlaybackDuration;
        }
    }
}

File: src/model/Voice.ts
import type { Bar } from '@src/model/Bar';
import type { Beat } from '@src/model/Beat';
import { GraceType } from '@src/model/GraceType';
import type { Settings } from '@src/Settings';
import { GraceGroup } from '@src/model/GraceGroup';
import { ElementStyle } from '@src/model/ElementStyle';

/**
 * Lists all graphical sub elements within a {@link Voice} which can be styled via {@link Voice.style}
 */
export enum VoiceSubElement {
    /**
     * All general glyphs (like notes heads and rests).
     */
    Glyphs = 0
}

/**
 * Defines the custom styles for voices.
 * @json
 * @json_strict
 */
export class VoiceStyle extends ElementStyle<VoiceSubElement> {}

/**
 * A voice represents a group of beats
 * that can be played during a bar.
 * @json
 * @json_strict
 */
export class Voice {
    private _beatLookup!: Map<number, Beat>;
    private _isEmpty: boolean = true;
    private _isRestOnly: boolean = true;

    private static _globalVoiceId: number = 0;

    /**
     * @internal
     */
    public static resetIds() {
        Voice._globalVoiceId = 0;
    }

    /**
     * Gets or sets the unique id of this bar.
     */
    public id: number = Voice._globalVoiceId++;

    /**
     * Gets or sets the zero-based index of this voice within the bar.
     * @json_ignore
     */
    public index: number = 0;

    /**
     * Gets or sets the reference to the bar this voice belongs to.
     * @json_ignore
     */
    public bar!: Bar;

    /**
     * Gets or sets the list of beats contained in this voice.
     * @json_add addBeat
     */
    public beats: Beat[] = [];

    /**
     * Gets or sets a value indicating whether this voice is empty.
     */
    public get isEmpty(): boolean {
        return this._isEmpty;
    }

    /**
     * The style customizations for this item.
     */
    public style?: VoiceStyle;

    /**
     * @internal
     */
    public forceNonEmpty() {
        this._isEmpty = false;
    }

    /**
     * Gets or sets a value indicating whether this voice is empty.
     */
    public get isRestOnly() {
        return this._isRestOnly;
    }

    public insertBeat(after: Beat, newBeat: Beat): void {
        newBeat.nextBeat = after.nextBeat;
        if (newBeat.nextBeat) {
            newBeat.nextBeat.previousBeat = newBeat;
        }
        newBeat.previousBeat = after;
        newBeat.voice = this;
        after.nextBeat = newBeat;
        this.beats.splice(after.index + 1, 0, newBeat);
    }

    public addBeat(beat: Beat): void {
        beat.voice = this;
        beat.index = this.beats.length;
        this.beats.push(beat);
        if (!beat.isEmpty) {
            this._isEmpty = false;
        }
        if (!beat.isRest) {
            this._isRestOnly = false;
        }
    }

    private chain(beat: Beat, sharedDataBag: Map<string, unknown> | null = null): void {
        if (!this.bar) {
            return;
        }
        if (beat.index < this.beats.length - 1) {
            beat.nextBeat = this.beats[beat.index + 1];
            beat.nextBeat.previousBeat = beat;
        } else if (beat.isLastOfVoice && beat.voice.bar.nextBar) {
            const nextVoice: Voice = this.bar.nextBar!.voices[this.index];
            if (nextVoice.beats.length > 0) {
                beat.nextBeat = nextVoice.beats[0];
                beat.nextBeat.previousBeat = beat;
            } else {
                beat.nextBeat!.previousBeat = beat;
            }
        }

        beat.chain(sharedDataBag);
    }

    public addGraceBeat(beat: Beat): void {
        if (this.beats.length === 0) {
            this.addBeat(beat);
            return;
        }
        // remove last beat
        const lastBeat: Beat = this.beats[this.beats.length - 1];
        this.beats.splice(this.beats.length - 1, 1);
        // insert grace beat
        this.addBeat(beat);
        // reinsert last beat
        this.addBeat(lastBeat);
        this._isEmpty = false;
        this._isRestOnly = false;
    }

    public getBeatAtPlaybackStart(playbackStart: number): Beat | null {
        if (this._beatLookup.has(playbackStart)) {
            return this._beatLookup.get(playbackStart)!;
        }
        return null;
    }

    public finish(settings: Settings, sharedDataBag: Map<string, unknown> | null = null): void {
        this._isEmpty = true;
        this._isRestOnly = true;
        this._beatLookup = new Map<number, Beat>();
        let currentGraceGroup: GraceGroup | null = null;
        for (let index: number = 0; index < this.beats.length; index++) {
            const beat: Beat = this.beats[index];
            beat.index = index;
            this.chain(beat, sharedDataBag);
            if (beat.graceType === GraceType.None) {
                beat.graceGroup = currentGraceGroup;
                if (currentGraceGroup) {
                    currentGraceGroup.isComplete = true;
                }
                currentGraceGroup = null;
            } else {
                if (!currentGraceGroup) {
                    currentGraceGroup = new GraceGroup();
                }
                currentGraceGroup.addBeat(beat);
            }
            if (!beat.isEmpty) {
                this._isEmpty = false;
            }
            if (!beat.isRest) {
                this._isRestOnly = false;
            }
        }

        let currentDisplayTick: number = 0;
        let currentPlaybackTick: number = 0;
        for (let i: number = 0; i < this.beats.length; i++) {
            const beat: Beat = this.beats[i];
            beat.index = i;
            beat.finish(settings, sharedDataBag);

            // if this beat is a non-grace but has grace notes
            // we need to first steal the duration from the right beat
            // and place the grace beats correctly
            if (beat.graceType === GraceType.None) {
                if (beat.graceGroup) {
                    const firstGraceBeat = beat.graceGroup!.beats[0];
                    const lastGraceBeat = beat.graceGroup!.beats[beat.graceGroup!.beats.length - 1];
                    if (firstGraceBeat.graceType !== GraceType.BendGrace) {
                        // find out the stolen duration first
                        const stolenDuration: number =
                            lastGraceBeat.playbackStart + lastGraceBeat.playbackDuration - firstGraceBeat.playbackStart;

                        switch (firstGraceBeat.graceType) {
                            case GraceType.BeforeBeat:
                                // steal duration from previous beat and then place grace beats newly
                                if (firstGraceBeat.previousBeat) {
                                    firstGraceBeat.previousBeat.playbackDuration -= stolenDuration;
                                    // place beats starting after new beat end
                                    if (firstGraceBeat.previousBeat.voice === this) {
                                        currentPlaybackTick =
                                            firstGraceBeat.previousBeat.playbackStart +
                                            firstGraceBeat.previousBeat.playbackDuration;
                                    } else {
                                        // stealing into the previous bar
                                        currentPlaybackTick = -stolenDuration;
                                    }
                                } else {
                                    // before-beat on start is somehow not possible as it causes negative ticks
                                    currentPlaybackTick = -stolenDuration;
                                }

                                for (const graceBeat of beat.graceGroup!.beats) {
                                    this._beatLookup.delete(graceBeat.playbackStart);
                                    graceBeat.playbackStart = currentPlaybackTick;
                                    this._beatLookup.set(graceBeat.playbackStart, beat);
                                    currentPlaybackTick += graceBeat.playbackDuration;
                                }

                                break;
                            case GraceType.OnBeat:
                                // steal duration from current beat
                                beat.playbackDuration -= stolenDuration;
                                if (lastGraceBeat.voice === this) {
                                    // with changed durations, update current position to be after the last grace beat
                                    currentPlaybackTick = lastGraceBeat.playbackStart + lastGraceBeat.playbackDuration;
                                } else {
                                    // if last grace beat is on the previous bar, we shift the time back to have the note played earlier
                                    currentPlaybackTick = -stolenDuration;
                                }
                                break;
                        }
                    }
                }

                beat.displayStart = currentDisplayTick;
                beat.playbackStart = currentPlaybackTick;

                if (beat.fermata) {
                    this.bar.masterBar.addFermata(beat.playbackStart, beat.fermata);
                } else {
                    beat.fermata = this.bar.masterBar.getFermata(beat);
                }

                this._beatLookup.set(beat.playbackStart, beat);
            } else {
                beat.displayStart = currentDisplayTick;
                beat.playbackStart = currentPlaybackTick;
            }

            beat.finishTuplet();
            if (beat.graceGroup) {
                beat.graceGroup.finish();
            }
            currentDisplayTick += beat.displayDuration;
            currentPlaybackTick += beat.playbackDuration;
        }
    }

    public calculateDuration(): number {
        if (this.isEmpty || this.beats.length === 0) {
            return 0;
        }
        const lastBeat: Beat = this.beats[this.beats.length - 1];
        const firstBeat: Beat = this.beats[0];
        return lastBeat.playbackStart + lastBeat.playbackDuration - firstBeat.playbackStart;
    }
}
